extensions/resume.md

```diff
@@ -0,0 +1,233 @@
+
+#### `RESUME SUCCESS` Message
+
+    RESUME SUCCESS <oldnick>
+
+The `SUCCESS` message indicates that the attempt to resume the connection from `<oldnick>` was successful.
+
+#### `RESUME WARN` Message
+
+    RESUME WARN <oldnick> <info>
+
+The `WARN` message indicates that there was a warning about the resume attempt, but not that it has succeeded or failed (it will be followed up by a `SUCCESS` or `ERR`). `<info>` is an information text string indicating the reson for the warning (e.g.: `"Chat history may be truncated"`).
+
+#### `RESUME ERR` Message
+
+    RESUME ERR <oldnick> <info>
+
+The `ERR` message indicates that the attempt to resume the connection from `<oldnick>` failed. `<info>` is an information text string indicating the reson for the failure (e.g.: `"Cannot resume connection, token is not valid"`).
+
+#### `RESUMED` Message
+
+    :nick!olduser@oldhost RESUMED <user> <host> [timestamp]
+
+This message is sent to indicate that a client has reconnected. `<nick>`, `<olduser>`, and `<oldhost>` indicate the client that has reconnected, and are the details of the old client. `<user>` and `<host>` indicate the reconnecting client's new username and hostname, and the client MUST process this information as they would a regular [`CHGHOST`](https://ircv3.net/specs/extensions/chghost-3.2.html) message. `<timestamp>`, if given, is a timestamp of the form described above which indicates the last message received by the reconnecting clent.
+
+Upon receiving a `RESUMED` message, clients MUST display in some way that the given user has reconnected. If `<timestamp>` is given, clients SHOULD use this to display how much message history seems to have been lost.
+
+
+## Capability Negotiation
+
+The first step in successfully resuming a connection is for the 'old client' to negotiate the `draft/resume-0.2` capability and receive a resumption token. This process is illustrated here:
+
+    C: CAP LS
+    S: CAP * LS :multi-prefix draft/resume-0.2
+    C: CAP REQ :draft/resume-0.2
+    S: CAP * ACK :draft/resume-0.2
+    S: RESUME TOKEN JKbAypzFiovffzuD8VEfcs6bOLrXsSenxsyZNt87Bb0YaQlyOzuQciz6cB2R
+
+
+## Resuming A Connection
+
+When a client detects that it has become disconnected from a server, it SHOULD try to resume before it breaks the existing connection.
+
+Upon establishing the new connection, the client begins capability negotiation, negotiates supported capabilities, and MUST confirm that the `draft/resume-0.2` capability exists. If this capability does not exist, the client continues connection registration without attempting to resume. If this capability does exist, the client sends the `RESUME` command and MUST wait for either a `RESUME SUCCESS` or a `RESUME ERR` message before continuing registration. It should be noted that the client MUST NOT perform SASL authentication if the `draft/resume-0.2` capability exists, as this will end connection registration and abort the resumption attempt.
```

Comment:
When implementing the server side of this (oragono#304), I got confused about whether the server needs to wait for `CAP END` before processing the resume token. The correct behavior is that the token should be processed immediately, i.e., a successful resume ends capability negotiation the same way `CAP END` does. This could probably use some clarifying language in the specification.

Reply:
Now clarified in the spec, thanks for the yell here.